mpsl: subsys: Add support for nrf21540_gpio and sky66112-11

AdamZelikNS opened the pull request:
User has the ability to use front-end module (FEM) type by
adding a node with label nrf_radio_fem to the devicetree.
When mpsl library is initialized selected FEM is configured
and becomes ready to use by higher layer protocol drivers.

Signed-off-by: Andrzej Kuros <andrzej.kuros@nordicsemi.no>

greg-fer commented:
@AdamZelikNS , I'll be out on Monday and Tuesday (Sept 14-15), and then on Friday, Sept 18. I'll try to have a look at this PR on Wednesday, Sept 16, after talking to @hubertmis .

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/nrfconnect/sdk-nrf?pullRequest=2896) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/nrfconnect/sdk-nrf?pullRequest=2896) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: ankuns<br/>:x: czeslawmakarski<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/nrfconnect/sdk-nrf?pullRequest=2896) it.</sub>

JakubPegza commented on subsys/mpsl/mpsl_init.c:
```diff
@@ -118,6 +121,125 @@ static uint8_t m_config_clock_source_get(void)
+	err = fem_nrf21540_gpio_configure();
+#elif CONFIG_MPSL_FEM_SKY66112_11
+	err = fem_sky66112_11_configure();
+#elif CONFIG_MSPL_FEM
```
typo: `MSPL`

AdamZelikNS replied:
fixed, thanks

greg-fer commented:
The page looks OK for me now, but the link in .rst to nrfxlib (:ref:`nrfxlib:mpsl_fem`) requires previous merge of updates from the Dragoon repo to nrfxlib. I'll review and approve once that is done.

AdamZelikNS commented:
I have rebased and removed commit "mpsl: subsys: Deny use of BLE with nrf21540_gpio" by @czeslawmakarski .

@greg-fer mpsl with fem has been merged into nrfxlib so you can continue reviewing.

hubertmis commented on doc/nrf/ug_radio_fem.rst:
```diff
@@ -0,0 +1,178 @@
+
+These methods are only available to protocol drivers that are using FEM features provided by the MPSL library in multiprotocol scenarios.
+They are also valid for cases where an application uses just one protocol, but benefits from features provided by MPSL.
+To avoid conflicts, check the protocol documentation to see if it uses FEM support provided by MPSL.
```
Shall we add a note here that currently none of protocols delivered by Nordic uses FEM support provided by MPSL, but it may be used by radio protocols implemented by customers?

hubertmis replied:
@czeslawmakarski , I think we briefly discussed it today. Do you have any comments?

czeslawmakarski replied:
I think some note like 'Currently no protocols support it' shall be enough.

AdamZelikNS replied:
I added such a note

hubertmis replied:
This note will be obsolete after next upmerge or hal_nordic revision update :)

hubertmis commented on subsys/mpsl/mpsl_init.c:
```diff
@@ -118,6 +121,125 @@ static uint8_t m_config_clock_source_get(void)
 #endif
 }
 
+#if CONFIG_MPSL_FEM_NRF21540_GPIO
```
I think you should use `#ifdef` for `bool` options from Kconfig

AdamZelikNS replied:
Good point. I used `#if IS_ENABLED(.)` for `bool`

hubertmis reviewed (commented):
Very good documentation and neat implementation :+1: 

I think we should add matching entries in dts of nRF21540-DK: https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/nrf21540dk_nrf52840/nrf21540dk_nrf52840.dts and show it as an example of correct dts for nRF21540-GPIO. I would wait with this merge until nRF21540-DK dts contains appropriate entries.

rugeGerritsen commented on subsys/mpsl/Kconfig:
```diff
@@ -25,6 +25,85 @@ config MPSL_TIMESLOT_SESSION_COUNT
+DT_COMPAT_NORDIC_NRF21540_GPIO := nordic,nrf21540_gpio
+DT_COMPAT_SKYWORKS_SKY66112_11 := skyworks,sky66112-11
+
+config MPSL_FEM
```
Is it supported on nrf53 already? If no, we should prevent the user from selecting this configuration

AdamZelikNS replied:
I added `depends on !(SOC_NRF5340_CPUAPP || SOC_NRF5340_CPUNET)`

joerchan replied:
This is temporary, right?

AdamZelikNS replied:
Yes, support for FEMs for nRF53 is on our roadmap, so from this point of view this `depends on !SOC_NRF53 ..` is temporary

rugeGerritsen reviewed (changes requested):
We need to prevent users from enabling the FEMs when Bluetooth LE is enabled

hubertmis commented on subsys/mpsl/Kconfig:
```diff
@@ -25,6 +25,85 @@ config MPSL_TIMESLOT_SESSION_COUNT
+	bool "Radio front-end module (FEM) support"
+	default $(dt_nodelabel_has_compat,nrf_radio_fem,$(DT_COMPAT_NORDIC_NRF21540_GPIO)) || \
+			$(dt_nodelabel_has_compat,nrf_radio_fem,$(DT_COMPAT_SKYWORKS_SKY66112_11))
+	help
```
It think we cannot do that. As long as flash driver compatible with MPSL depends on `BT_CTLR` we cannot remove `BT_CTLR` from build when using PA/LNA from MPSL.
After adding flash driver compatible with MPSL to Zephyr tree (I think this work is in progress), we could add such negative dependency to Kconfig

czeslawmakarski replied:
This is a bit more complicated, since if you want to have MPSL plus Flash access, then you have to have softdevice controller linked in. If we explicitl deny it as you suggested, then no customer shall be able to make a product with FEM and Flash access.

rugeGerritsen replied:
Thanks for providing this insight. I understand that we have to leave it as is for now.. If you're not aware of it, there is already a draft PR open to resolve this odd dependency: https://github.com/tejlmand/fw-nrfconnect-nrf/pull/2

AdamZelikNS replied:
I added `depends on !BT` in `config MPSL_FEM_NRF21540_GPIO` only. Tf I am not mistaken then there is no such requirement for `SKY66112` because it has a bypass feature, so it will not block BLE even if not properly controlled. Do you think its ok or should this apply to all types of FEM after all?

joerchan replied:
@rugeGerritsen Is this OK?

rugeGerritsen replied:
Unless there is a use case where you would like to have FEM for 802.15.4 but not for Bluetooth LE, I would rather add `depends on !BT` to make it clear that it is not yet supported

czeslawmakarski commented on subsys/mpsl/mpsl_init.c:
```diff
@@ -118,6 +121,125 @@ static uint8_t m_config_clock_source_get(void)
+				DT_PROP(DT_NODELABEL(nrf_radio_fem), rx_gain_db)
+		},
+		.pa_pin_config = {
+			.enable       = true,
```
In case of nRF21540 GPIO interface we are targeting one specific FEM (from Nordic, nRF21540), so it is not meant to be 'generic' or something, which is not the case in case of 'Simple GPIO' FEM. nRF21540 has a negative PDN polarity, so we this is exactly what we provide.

rugeGerritsen replied:
Are there cases were the user only wants to enable the PA or only the LNA? If not, I'm fine with the configuration

AdamZelikNS replied:
Good suggestion. It may be advantageous if these two parameters are also configurable with device-tree. I will add one more commit with this

AdamZelikNS replied:
Now it is possible to set the polarity and also to disable pins

greg-fer commented:
Is this PR going into NCS 1.4.0?

AdamZelikNS commented:
No, we don't really need this in 1.4.0.  It can be merged later.

KAGA164 commented:
It think it would be useful if the nrf21540 or other FEM devices could be used without whole MPSL initialization. We have some samples like the DTM or the Radio Test when we need a driver for the nrf21540 without give up control over RADIO, TIMER ... to the MPSL.

joerchan commented:
> It think it would be useful if the nrf21540 or other FEM devices could be used without whole MPSL initialization. We have some samples like the DTM or the Radio Test when we need a driver for the nrf21540 without give up control over RADIO, TIMER ... to the MPSL.

@KAGA164 You still have control over the RADIO and TIMER peripheral if you use timeslot. I think it would be best to use MPSL instead of creating MPSL-less drivers in addition.

KAGA164 commented:
> @KAGA164 You still have control over the RADIO and TIMER peripheral if you use timeslot. I think it would be best to use MPSL instead of creating MPSL-less drivers in addition.

@joerchan I`m not talking about making a new driver. The FEM API can be used without initializing whole MPSL. This is possible now when you remove adding mpsl_init.c source file from application but enabling MPLS always include this file.

AdamZelikNS commented:
Made most of the suggested fixes. Rebased.

AdamZelikNS commented on subsys/mpsl/mpsl_init.c:
```diff
@@ -155,6 +277,11 @@ static int mpsl_lib_init(struct device *dev)
 	IRQ_DIRECT_CONNECT(RADIO_IRQn, MPSL_HIGH_IRQ_PRIORITY,
 			   mpsl_radio_isr_wrapper, IRQ_ZERO_LATENCY);
 
+	err = fem_configure();
```
I added it. Agree, it looks more readable

AdamZelikNS commented on subsys/mpsl/mpsl_init.c:
```diff
@@ -118,6 +121,125 @@ static uint8_t m_config_clock_source_get(void)
+#if CONFIG_MPSL_FEM_NRF21540_GPIO
+static int fem_nrf21540_gpio_configure(void)
+{
+	/* FEM configuration requires gpiote and ppi channels.
```
Good point. Limitation apply to both. Copied to `fem_sky66112_11_configure`

AdamZelikNS commented on dts/bindings/radio_fem/skyworks,sky66112-11.yaml:
```diff
@@ -0,0 +1,48 @@
+include: base.yaml
+
+properties:
+    ctx-pin:
```
I added a mention of naming conventions in the .rst

KAGA164 commented on subsys/mpsl/Kconfig:
```diff
@@ -25,6 +25,95 @@ config MPSL_TIMESLOT_SESSION_COUNT
+
+if MPSL_FEM_NRF21540_GPIO
+
+config MPSL_FEM_NRF21540_GPIO_GPIOTE_TX_EN
```
Is there any reason why GPIOTE and PPI configuration is in the Kconfig instead of in the DTS ?

hubertmis replied:
GPIOTE and PPI shall be taken from software allocators. Right now only allocator for PPI is available. GPIOTE is in progress. When it is implemented, we shall update this module to remove GPIOTE and PPI allocation from Kconfig and use automatic software allocators instead.

KAGA164 commented on dts/bindings/radio_fem/skyworks,sky66112-11.yaml:
```diff
@@ -0,0 +1,48 @@
+            Pin number of the SOC controlling CTX pin of the SKY66112-11
+
+    crx-pin:
+        type: int
```
Wouldn't it be better to use here a `gpios` `phandle-array` type since you need for example information about pin polarity ?

hurricaneFromMoscow commented on doc/nrf/links.txt:
```diff
@@ -89,6 +89,10 @@
 
 .. _`nRF52840 Product Specification`: https://infocenter.nordicsemi.com/topic/ps_nrf52840/keyfeatures_html5.html
 
+.. _`nRF21540 Objective Product Specification`: https://infocenter.nordicsemi.com/topic/ps_nrf21540/keyfeatures_html5.html
```
This link looks broken.

hubertmis replied:
```suggestion
.. _`nRF21540 Product Specification`: https://infocenter.nordicsemi.com/topic/struct_fem/struct/nrf21540_ps.html
```

AdamZelikNS replied:
I fixed the link as suggested

hubertmis commented on doc/nrf/ug_radio_fem.rst:
```diff
@@ -0,0 +1,190 @@
+
+  .. note::
+    These values have some constraints.
+    For details, see `nRF21540 Objective Product Specification`_.
```
```suggestion
    For details, see `nRF21540 Product Specification`_.
```

AdamZelikNS replied:
I applied it

hubertmis reviewed (commented):
It seems some changes are still pending. Because of that I don't approve yet.
@FrancescoSer, could you have a look at links update I've suggested? It looks there is not more nRF21540 OPS in inforcenter. I've replaced the link with nRF21540 PS.

grochu commented on subsys/mpsl/mpsl_init.c:
```diff
@@ -124,6 +128,171 @@ static uint8_t m_config_clock_source_get(void)
+		},
+		.pdn_pin_config = {
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), pdn_gpios)
+			.enable       = true,
```
Can the PDN management be made configurable? E.g. in the nRF21540 "mixed-mode" control, where PA/LNA is controlled from GPIOs and some configuration fields such as TX_GAIN are set through SPI, it will be the application who will control the PDN to avoid clearing the configured settings after each Radio transaction. This will be the case in the RF testing samples such DTM.

AdamZelikNS replied:
Assuming the described feature request for PDN is valid only together with nRF21540-mixed-mode avaliability - 
I can see it this way that with nRF21540-mixed-mode a new separate FEM initialization method will be added in mpsl_init.c. Then a new variant of the PDN pin configuration will be taken into account, most likely using an additional flag from the Kconfig.

grochu replied:
Agree, this is a good approach.

hubertmis replied:
But I think such additional mode could be implemented later, out of scope of this PR. Right?

AdamZelikNS replied:
Yes, I meant adding "mixed-mode" and a new way to control pin PDN in the next step in the future so it is beyond the scope of this PR.

lukasbrchl commented:
Hi, would it be possible to use nRF21540 in mixed mode? Setting TX_GAIN using SPI but toggling RX/TX by GPIO? There is an issue that whenever you go to the Program mode, the TX_GAIN resets. See the last comment from Andreas - https://devzone.nordicsemi.com/f/nordic-q-a/66978/the-nrf21540-amplifier-can-t-enable-transmit-mode-after-receive-mode/274334?

hubertmis commented:
@lukasbrchl , it seems to me that this PR is not about power control, it's only covering timings of PDN, TX_EN, and RX_EN pins. This module could be enhanced in the future with managing the MODE pin for power control.

Running mixed mode like you described should be implemented in a separated module. @KAGA164 could knows more details about such module for mixed mode.

lukasbrchl commented:
Ok, I thought this is about implementing nrf21540 driver for NCS. And I would expect that people will use it as I described - SPI for TX_GAIN and GPIO for rx/tx control. Thats why I think it is relevant.

hubertmis commented:
> I thought this is about implementing nrf21540 driver for NCS

That's correct. However it is implemented iteratively. The first step is just timing control with PDN, RX_EN and TX_EN (covered in this PR). Power control will be implemented later.

joerchan commented:
@mbolivar-nordic  and @tejlmand Could you have a look at the Integration of Kconfig and Devicetree?

joerchan commented on subsys/mpsl/Kconfig:
```diff
@@ -25,6 +25,108 @@ config MPSL_TIMESLOT_SESSION_COUNT
+	default MPSL_FEM_SKY66112_11 if CAN_USE_MPSL_FEM_SKY66112_11
+
+config MPSL_FEM_NRF21540_GPIO
+	depends on !BT
```
I think it is better to do: `depends on MPSL_FEM_NRF2154_GPIO_SUPPORT`

joerchan commented on subsys/mpsl/Kconfig:
```diff
@@ -25,6 +25,108 @@ config MPSL_TIMESLOT_SESSION_COUNT
+	  nRF21540 is unavailable when BLE is enabled
+
+config MPSL_FEM_SKY66112_11
+	bool "SKY66112-11 front-end module"
```
Similar: `depends on MPSL_FEM_SKY66112_11_SUPPORT`

joerchan commented on subsys/mpsl/Kconfig:
```diff
@@ -25,6 +25,108 @@ config MPSL_TIMESLOT_SESSION_COUNT
+
+choice MPSL_FEM_CHOICE
+	prompt "Radio front-end module (FEM) type"
+	default MPSL_FEM_NRF21540_GPIO if CAN_USE_MPSL_FEM_NRF21540_GPIO
```
Instead of putting the defaults here, order the choices in the order that is preferred, so that when both are supported you end up with default choice being `MPSL_FEM_NRF21540_GPIO ` (which I assume is what we want?)

AdamZelikNS replied:
That's right, without these defaults it also works as designed.

joerchan commented on subsys/mpsl/Kconfig:
```diff
@@ -25,6 +25,108 @@ config MPSL_TIMESLOT_SESSION_COUNT
+	depends on !BT
+	bool "nRF21540 front-end module in GPIO mode"
+	help
+	  nRF21540 is unavailable when BLE is enabled
```
This help text is a bit short IMO.
Also, don't use the abbreviation BLE, use, Bluetooth LE instead (or Low Energy).

AdamZelikNS replied:
My intention is to post a note here emphasizing only one thing - this relationship:
`nRF21540 is unavailable when Bluetooth LE is enabled`
Apart from that, I don't think any general description is needed.

joerchan replied:
Fair enough.

joerchan commented on subsys/mpsl/Kconfig:
```diff
@@ -25,6 +25,108 @@ config MPSL_TIMESLOT_SESSION_COUNT
+config MPSL_FEM
+	bool "Radio front-end module (FEM) support"
+# MPSL FEM is not supported on 53 yet
+	depends on !(SOC_NRF5340_CPUAPP || SOC_NRF5340_CPUNET)
```
`depends on !(...) && (MPSL_FEM_NRF2154_GPIO_SUPPORT || MPSL_FEM_SKY66112_11_SUPPORT)`

joerchan commented on subsys/mpsl/Kconfig:
```diff
@@ -25,6 +25,108 @@ config MPSL_TIMESLOT_SESSION_COUNT
+	default CAN_USE_MPSL_FEM_NRF21540_GPIO || CAN_USE_MPSL_FEM_SKY66112_11
+	help
+	  Controls if front-end module (FEM) it to be configured automatically
+	  when MPSL is initialized. Default value and type of FEM to use
```
Defaults should be in Kconfig (and as far as I can see they are).

AdamZelikNS replied:
I changed that description a bit

mbolivar-nordic commented:
> @mbolivar-nordic and @tejlmand Could you have a look at the Integration of Kconfig and Devicetree?

I didn't have time to look at this in detail today and I'm on holiday until next week now. I will check on Monday.

KAGA164 commented on dts/bindings/radio_fem/nordic,nrf21540-fem.yaml:
```diff
@@ -0,0 +1,70 @@
+# Copyright (c) 2020 Nordic Semiconductor ASA
```
Isn't that a duplication of that https://github.com/zephyrproject-rtos/zephyr/commit/19a93a3c68c79dd920678f38fbfcfd66fe32afee ?

AdamZelikNS replied:
Yes. Good point. I remove it from this pr

KAGA164 commented on subsys/mpsl/mpsl_init.c:
```diff
@@ -124,6 +128,171 @@ static uint8_t m_config_clock_source_get(void)
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), tx_en_gpios)
+			.enable       = true,
+			.active_high  = MPSL_FEM_NRF21540_GPIO_POLARITY_GET(tx_en_gpios),
+			.gpio_pin     = DT_GPIO_PIN(DT_NODELABEL(nrf_radio_fem), tx_en_gpios),
```
What if the device has more than one GPIO port ?
For example:
`DT_GPIO_PIN` for `<&gpio0 12 GPIO_ACTIVE_LOW>` and `<&gpio1 12 GPIO_ACTIVE_LOW>` is 12 in both cases

AdamZelikNS replied:
I fixed the initialization - previously it was not possible to use GPIO-1 pins, at least not in a clean way.
Now you can use pins from both `&gpio0` and `&gpio1` in any combination.

AdamZelikNS commented on subsys/mpsl/Kconfig:
```diff
@@ -25,6 +25,108 @@ config MPSL_TIMESLOT_SESSION_COUNT
+DT_COMPAT_NORDIC_NRF21540_GPIO := nordic,nrf21540-fem
+DT_COMPAT_SKYWORKS_SKY66112_11 := skyworks,sky66112-11
+
+config CAN_USE_MPSL_FEM_NRF21540_GPIO
```
Yes, FEM has nothing to do with CAN :)
Renamed

hubertmis commented on subsys/mpsl/Kconfig:
```diff
@@ -25,6 +25,108 @@ config MPSL_TIMESLOT_SESSION_COUNT
+	bool "Radio front-end module (FEM) support"
+# MPSL FEM is not supported on 53 yet
+	depends on !(SOC_NRF5340_CPUAPP || SOC_NRF5340_CPUNET)
+	default CAN_USE_MPSL_FEM_NRF21540_GPIO || CAN_USE_MPSL_FEM_SKY66112_11
```
@joerchan , I would like to understand the reasons to remove default.

If nRF21540 is present in the design, it must be enabled in software. If it is disabled, radio protocols are unusable, it would work as an attenuator.

I thought that options in Kconfig should depend on devicetree description.
1. If nRF21540 is missing, it cannot be enabled in Kconfig.
2. If nRF21540 GPIO interface is defined, it is the only available option in Kconfig (and default one).
3. If nRF21540 SPI interface is defined, it is the only one available option in Kconfig
4. If both GPIO and SPI interfaces are defined there are three options in Kconfig (SPI is default):
    1. SPI
    2. GPIO
    3. Mixed (SPI for power, GPIO for timings)

Currently only GPIO binding is defined (and implemented). Because of that I was expecting support for points 1 and 2 from the list above in this PR.

What are your opinions?

joerchan replied:
@hubertmis I meant that the default selected based on the support. Because the support should go with depends instead as they are now.

Having `default y` is fine.

hubertmis replied:
@joerchan , do I understand correctly that you suggest something like:
```
config MPSL_FEM
	bool "Radio front-end module (FEM) support"
	depends on !(SOC_NRF5340_CPUAPP || SOC_NRF5340_CPUNET) && \
	           (MPSL_FEM_NRF21540_GPIO_SUPPORT || MPSL_FEM_SKY66112_11_SUPPORT)
        default y

choice MPSL_FEM_CHOICE
	prompt "Radio front-end module (FEM) type"

#This new option will be added after SPI support is implemented
config MPSL_FEM_NRF21540_SPI
	depends on MPSL_FEM_NRF21540_SPI_SUPPORT
	bool "nRF21540 front-end module in SPI mode"
        default y

config MPSL_FEM_NRF21540_GPIO
	depends on MPSL_FEM_NRF21540_GPIO_SUPPORT
	bool "nRF21540 front-end module in GPIO mode"
        default y if !MPSL_FEM_NRF21540_SPI_SUPPORT # To be added after SPI support is added
	help
	  Note: nRF21540 is unavailable when Bluetooth LE is enabled

#This new option will be added after SPI support is implemented
config MPSL_FEM_NRF21540_MIXED
	depends on MPSL_FEM_NRF21540_SPI_SUPPORT && MPSL_FEM_NRF21540_GPIO_SUPPORT
	bool "nRF21540 front-end module in mixed mode"
	help
	  Mixed mode means that SPI is used to control TX power, while GPIO is used to control radio timings. This mode should not be used for low-power applications, because nRF21540 is powered on all the time to keep TX power settings in its ram. For low-power applications use MPSL_FEM_NRF21540_SPI instead.

config MPSL_FEM_SKY66112_11
	depends on MPSL_FEM_SKY66112_11_SUPPORT
	bool "SKY66112-11 front-end module"
```

joerchan replied:
@hubertmis As i mentioned in another of the comments the order of choices determines the defaults, so it is better to order them in preferred order to avoid putting in too many default statements with condititions.

So like this I think would be best:
```
config MPSL_FEM
	bool "Radio front-end module (FEM) support"
	depends on !(SOC_NRF5340_CPUAPP || SOC_NRF5340_CPUNET) && \
	           (MPSL_FEM_NRF21540_GPIO_SUPPORT || MPSL_FEM_SKY66112_11_SUPPORT)
        default y

choice MPSL_FEM_CHOICE
	prompt "Radio front-end module (FEM) type"

#This new option will be added after SPI support is implemented
config MPSL_FEM_NRF21540_SPI
	depends on MPSL_FEM_NRF21540_SPI_SUPPORT
	bool "nRF21540 front-end module in SPI mode"

config MPSL_FEM_NRF21540_GPIO
	depends on MPSL_FEM_NRF21540_GPIO_SUPPORT
	bool "nRF21540 front-end module in GPIO mode"
	help
	  Note: nRF21540 is unavailable when Bluetooth LE is enabled

#This new option will be added after SPI support is implemented
config MPSL_FEM_NRF21540_MIXED
	depends on MPSL_FEM_NRF21540_SPI_SUPPORT && MPSL_FEM_NRF21540_GPIO_SUPPORT
	bool "nRF21540 front-end module in mixed mode"
	help
	  Mixed mode means that SPI is used to control TX power, while GPIO is used to control radio timings. This mode should not be used for low-power applications, because nRF21540 is powered on all the time to keep TX power settings in its ram. For low-power applications use MPSL_FEM_NRF21540_SPI instead.

config MPSL_FEM_SKY66112_11
	depends on MPSL_FEM_SKY66112_11_SUPPORT
	bool "SKY66112-11 front-end module"
```

rugeGerritsen commented on subsys/mpsl/mpsl_init.c:
```diff
@@ -127,6 +132,260 @@ static uint8_t m_config_clock_source_get(void)
+#endif
+		},
+		.ppi_channels = {
+			CONFIG_MPSL_FEM_NRF21540_GPIO_PPI_CHANNEL_0,
```
When allocating PPI channels, this must either be either be allocated run-time like done here: https://github.com/nrfconnect/sdk-nrf/blob/master/subsys/mpsl/mpsl_init.c#L137.
Or compile time, like done here: https://github.com/nrfconnect/sdk-nrf/blob/master/subsys/bluetooth/controller/hci_driver.c#L52
Otherwise the application may end up in a situation where a PPI channel is reconfigured.

Currently there is no nice way of doing it. @anangl FYI

hubertmis replied:
Discussed offline. We intend to keep current proposal temporarily, until thread-safe channel allocator for PPI and GPIOTE is available in NCS.

rugeGerritsen commented on subsys/mpsl/Kconfig:
```diff
@@ -25,6 +25,107 @@ config MPSL_TIMESLOT_SESSION_COUNT
 	help
 	  Maximum number of timeslot sessions.
 
+DT_COMPAT_NORDIC_NRF21540_GPIO := nordic,nrf21540-fem
```
I would consider moving the FEM configuration to a separate file: Kconfig.fem

tejlmand replied:
and maybe even consider if the settings specifically related to nRF21540 and sky66112 should be placed independent files as well.

Also, why are pin numbers defined in Kconfig and not in DTS ?

AdamZelikNS replied:
I made the split as suggested.

In my opinion, it is not necessary to separate nRF21540 and sky66112 now. I will consider this in the future when adding nRF21540-SPI and nRF21540-mixed. Then these files will probably be too extensive.

Regarding the pin numbers, @tejlmand, you probably meant that PPI and GPIOTE are defined in Kconfig.
I will refer to the explanations @hubertmis has posted [here](https://github.com/nrfconnect/sdk-nrf/pull/2896#discussion_r537591023) - we intend to keep current proposal temporarily.

rugeGerritsen commented on subsys/mpsl/mpsl_init.c:
```diff
@@ -127,6 +132,260 @@ static uint8_t m_config_clock_source_get(void)
 #endif
 }
 
+#if IS_ENABLED(CONFIG_MPSL_FEM)
```
I would consider moving the FEM configuration to a separate file: mpsl_fem.c

rugeGerritsen reviewed (changes requested):
PPI allocation must integrate with the nrfx ppi driver. Otherwise I just left some minor comments.

hubertmis reviewed (approved):
Looks good :+1: 
I agree that FEM-related parts could be moved to separated .c and Kconfig files. But it's not necessity from my point of view.

hubertmis commented:
@joerchan, I've approved this PR like an hour ago. Have you intentionally dismissed my approval?

hubertmis commented on subsys/mpsl/Kconfig:
```diff
@@ -25,6 +25,107 @@ config MPSL_TIMESLOT_SESSION_COUNT
+	bool "Radio front-end module (FEM) support"
+# MPSL FEM is not supported on 53 yet
+	depends on !(SOC_NRF5340_CPUAPP || SOC_NRF5340_CPUNET) && \
+	           (MPSL_FEM_NRF21540_GPIO_SUPPORT || MPSL_FEM_SKY66112_11_SUPPORT)
```
```suggestion
	           (MPSL_FEM_NRF21540_GPIO_SUPPORT || MPSL_FEM_SKY66112_11_SUPPORT)
	default y
```
It seems default is still missing

AdamZelikNS replied:
:thumbsup: 
I took this into account when I introduced a separate file "Kconfig.fem"

mbolivar-nordic commented on dts/bindings/radio_fem/skyworks,sky66112-11.yaml:
```diff
@@ -0,0 +1,46 @@
+
+    ctx-settle-time-us:
+        type: int
+        default: 23
+        description: >
+            Settling time in microseconds from activation of CTX to trasnsmit.
+
+    crx-settle-time-us:
+        type: int
+        default: 5
+        description: >
+            Settling time in microseconds from activation of CRX to receive.
+
+    tx-gain-db:
+        type: int
+        default: 22
+        description: >
+            TX gain of the PA amplifier of SKY66112-11 in dB.
+
+    rx-gain-db:
+        type: int
+        default: 11
+        description: >
+            RX gain of the LNA amplifier of SKY66112-11 in dB.
```
The zephyr policy is that all defaults must be documented with a justifiable rationale.

For example, if these are the chip's power-on-reset values, and they can be individually configured away from these defaults without affecting other options, the descriptions should say that those are the power-on-reset values.

But if the power on reset values must be configured in a group, then the policy is to make the properties all required.

For example, if I can't change `rx-gain-db` without also having to change `tx-gain-db`, then there should be no default values, and the properties should just be required, with an explanation of some recommended defaults in the top-level description for the binding (right after "SKY66112-11 Radio Front-End module.").

AdamZelikNS replied:
I added a description of defaults

mbolivar-nordic commented on dts/bindings/radio_fem/skyworks,sky66112-11.yaml:
```diff
@@ -0,0 +1,46 @@
+
+properties:
+    ctx-gpios:
+        type: phandle-array
```
Thank you for making these phandle-arrays here. That's better than `ctx-pin` values. We'd like to get away from `foo-pin` properties in general, even though we use them sometimes in Nordic IP blocks because it makes it more convenient to initialize PSEL registers. The phandle array is better because it refers to the actual GPIO controller by phandle, which the tooling can see.  Since the tools know a GPIO controller, this will enable future changes that let us build a pin map for the SoC in a generic way. The `foo-pin` type options don't let the tooling know that the property is actually configuring a pin.

mbolivar-nordic reviewed (changes requested):
Hi, I just took a look at the binding for now, since there are some ongoing discussions about where to add documentation for devicetree I will just let those alone.

Overall it looks great. The only reason I am putting a -1 on is because the default values need to be justified and documented in the binding.  Otherwise the binding seems very reasonable to me and I think it's good to go.

Thanks for your patience.

tejlmand commented on subsys/mpsl/Kconfig:
```diff
@@ -25,6 +25,107 @@ config MPSL_TIMESLOT_SESSION_COUNT
+	depends on MPSL_FEM_NRF21540_GPIO_SUPPORT
+	bool "nRF21540 front-end module in GPIO mode"
+	help
+	  Note: nRF21540 is unavailable when Bluetooth LE is enabled
```
This is a user configurable setting, so there should be a proper help text here to let the user understand what this is about.

tejlmand commented on subsys/mpsl/Kconfig:
```diff
@@ -25,6 +25,107 @@ config MPSL_TIMESLOT_SESSION_COUNT
+
+config MPSL_FEM_SKY66112_11
+	depends on MPSL_FEM_SKY66112_11_SUPPORT
+	bool "SKY66112-11 front-end module"
```
This is a user configurable setting, so there should be a proper help text here to let the user understand what this is about.

tejlmand commented on subsys/mpsl/Kconfig:
```diff
@@ -25,6 +25,107 @@ config MPSL_TIMESLOT_SESSION_COUNT
+
+config MPSL_FEM_NRF21540_GPIO_GPIOTE_TX_EN
+	int "GPIOTE channel number used to control TX_EN pin of the front-end module"
+	default 7
```
This is a user configurable setting, so there should be a proper help text here to let the user understand what this is about and how it might relate to other settings.

The prompt for this setting seems a bit long, consider if can be shortened and instead provide additional info in the help text.

AdamZelikNS replied:
I reworked prompts and added missing help parts in this file

mbolivar-nordic commented on dts/bindings/radio_fem/skyworks,sky66112-11.yaml:
```diff
@@ -0,0 +1,73 @@
+    ctx-settle-time-us:
+        type: int
+        default: 23
+        description: >
+            Settling time in microseconds from activation of CTX to transmit.
+
+            The default value has been determined experimentally: it is the
+            minimum time interval between asserting the CTX signal and starting
+            the radio transmission, which does not generate unacceptable
+            interference level on adjacent channels. The tests concerned a
+            configuration in which nRF5x SoC controlling the SKY66112-11.
+            The value can be adjusted for other tuned configurations.
+
+    crx-settle-time-us:
+        type: int
+        default: 5
+        description: >
+            Settling time in microseconds from activation of CRX to receive.
+
+            The default value has been determined experimentally: it is the
+            minimum time interval between asserting the CRX signal and starting
+            the radio reception, which does not degrade transmission quality
+            indicators (compared to a constantly turned on FEM LNA). The tests
+            concerned a configuration in which nRF5x SoC controlling
+            the SKY66112-11. The value can be adjusted for other tuned
+            configurations.
```
I'm a bit uncomfortable with the justifications for these defaults (for ctx-settle-time-us and crx-settle-time-us). It seems like they might not be following the rules, which are laid out here:

https://docs.zephyrproject.org/latest/guides/dts/bindings.html

When it says that it's been tuned for a particular configuration and may need to be adjusted for other configurations, it seems like they are potentially not independent, which would break this rule:

> If changing one property would require changing another to create a
> consistent configuration then those properties should be made
> required.

It also seems like it's just tuned for one SoC, which in general is not appropriate for defaults -- the consensus is that those should be made required, and recommended defaults for different situations should be documented.

hubertmis replied:
> it seems like they are potentially not independent

I think they are independent. I am aware of users that tuned only TX timing, because RX was correct for their application.


> It also seems like it's just tuned for one SoC

That's a valid point. It looks these parameters work correctly for the whole SoC family (at least nRF52). I'm just thinking, that we are adding this binding to sdk-nrf, which is intended only for nRF SoCs. Perhaps we could relax this rule and keep SoCs related default in binding to make creating DTSs easier for SDK users?

If not, I'm fine with making these parameters required and documenting recommended values.

mbolivar-nordic replied:
> I think they are independent. I am aware of users that tuned only TX timing, because RX was correct for their application.

OK, that resolves this concern. Thanks for confirming.

> I'm just thinking, that we are adding this binding to sdk-nrf, which is intended only for nRF SoCs. Perhaps we could relax this rule and keep SoCs related default in binding to make creating DTSs easier for SDK users?

Yeah, I thought about that too. The downsides of this approach are:

- If we ever upstream this binding, it is unlikely to be accepted with these defaults. So at that point, the defaults would probably switch to `required` and existing user applications would break. Might not be a real problem if that never happens.

- In general, I have observed that when NCS behaves inconsistently with zephyr, it causes significant user confusion. Upstream vs. downstream flash partitioning and multi-image builds are the main example here. So things like this generally make me nervous. Might not be a real problem here.

What about this compromise? The driver can use DT_NODE_HAS_PROP() to check if the relevant properties are defined.

- If they are defined, just use them from the DT using DT_PROP() (or the INST versions of those macros).

- If they are not defined, the driver emits a warning and uses the current defaults in the binding.

If that's not acceptable, I would defer to area maintainers on whether making an exception in this case is worth it.

hubertmis replied:
I think we should decide now how to handle that this driver is generic, what is discussed under gain comment. I'll get to this later.

mbolivar-nordic commented on dts/bindings/radio_fem/skyworks,sky66112-11.yaml:
```diff
@@ -0,0 +1,73 @@
+            TX gain of the PA amplifier of SKY66112-11 in dB.
+
+            Default value is based on Table 5 of the SKY66112-11 data sheet
+            (July 19, 2019). This is a fixed parameter of the device. The value
+            can be changed if, instead of SKY66112-11, another device with
+            different gain, but compatible with the control interface via the
+            CTX, CRX pins is used.
+
+    rx-gain-db:
+        type: int
+        default: 11
+        description: >
+            RX gain of the LNA amplifier of SKY66112-11 in dB.
+
+            Default value is based on Table 5 of the SKY66112-11 data sheet
+            (July 19, 2019). This is a fixed parameter of the device. The value
+            can be changed if, instead of SKY66112-11, another device with
+            different gain, but compatible with the control interface via the
+            CTX, CRX pins is used.
```
If these are truly fixed parameters of the device, I don't follow why they are here in the binding. The driver should be able to `#define` them based on the compatible itself. If you want to use "another device", that would require its own devicetree binding.

hubertmis replied:
The driver for sky66112 can be used with multiple similar front end modules. The modules differ only in required timings and gain values. It would be hard to maintain and to create another very similar binding and driver for each of nearly identical devices, that differs only in one `#define` (and potentially some timings defined in DTS).

What would be the correct way to provide a generic driver and matching binding, that can be used with sky66112 and similar front end modules?

mbolivar-nordic replied:
I have no experience with FEMs, but in general there are two options when defining bindings for multiple related devices:

1. create a YAML "include" file that does not define a compatible but defines the common properties, and include it from the "real" binding files for each actual device. An example "include" file is nordic,nrf-spi-common.yaml, which is included by the "real" binding files nordic,nrf-spim.yaml and nordic,nrf-spi.yaml.

2. create a generic binding that describes the entire category of devices and is supported by a common driver. I think jedec,spi-nor.yaml is an example.

hubertmis replied:
Let my explain our intentions here. We would like to add to NCS support for Front-End Modules like:

- SKY66112
- SKY66113 (using subset of features)
- SKY66114
- SKY66403
- SKY66404
- 8TR8201

Each of these devices have the same control interface. However, they differ in gain values and slightly in required timings. These are just examples, I don't think that I've found all FEMs sharing this simple generic GPIO interface. Becasue of that we would like to make this list as easily extensible as possible. I don't think that this simple interface (two GPIO pins, one used to indicate TX, another one used to indicate RX) is anywhere specified like Jedec SPI. Moreover, these devices sharing the same simple control interface are created by multiple vendors.

The driver for SKY66112 can be easily used for controlling other devices. Usually timings we have for SKY66112 apply to other. The only thing that should be adjusted is gain of particular device in given design (please note that gain may vary for the same device depending on RF path design on given board).

Our goal is to enable NCS users, that have one of FEMs mentioned above or similar in their design, to easily enable FEM support in NCS project. The easiest way is to define control pin numbers (done in device tree), define gains (if differ from the example we provided), and optionally tweak timings (usually not necessary).

I'm not sure which of the options you suggested applies to our intentions.

mbolivar-nordic replied:
Thanks for explaining. @AdamZelikNS and I discussed offline and I think we settled on using a generic include file with individual hardware bindings for real devices that mpsl_fem.c finds in the tree is the way to go.

We also discussed using a chosen node or an alias as a more flexible way to find the desired node. Adam will be looking into it further.

AdamZelikNS replied:
I came to the conclusion that using alias or chosen node not be of benefit in this case. This is due to the fact that there is most probably only one FEM instance in the system.

The approach that, as I applied, i.e. 
  - there is generic-fem-two-ctrl-pins DT binding
  - binding for SKY66112 supplements the generics one with device-specific values
  - driver (and its Kconfig) is generic as it are looking for `compatible = "generic-fem-two-ctrl-pins"` and using it

is well suited to our needs.

joerchan reviewed (changes requested):
Please squash the commits before merge this PR.
